### ISSUE TITLE
Alternate 2D Rendering method

### ISSUE THREAD

**Issue by fzurita** [gonetz/GLideN64#1282]
See this link:

https://www.libretro.com/index.php/introducing-vulkan-psx-renderer-for-beetlemednafen-psx/

TinyTiger (a very smart graphics API programmer) did his own PSX graphics plugin.

In his plugin he implemented his own post processor that somehow determines what is a 3D graphics and renders it at original resolution and anything 2D is downscaled to original resolution and filtered. He is a clever programmer. Source code is in the link, I will see if I can find the post processor and if it can be implemented in this plugin.

**Comment by fzurita** [comment]
He also describes a way to reordering draw calls to allow for batching them to help performance. His method would help performance a lot with Adreno.

**Comment by fzurita** [comment]
I found the glsl shader "source code" for the Adative Smoothing. Unfortunately, it looks like they decided not to release it and instead opted to release the shaders in a pre-compiled format. Here is the link anyways:

https://github.com/libretro/beetle-psx-libretro/blob/e4515c25e663ffe0bce3a18b43d080d9a64efd0f/parallel-psx/glsl/prebuilt/scaled.quad.frag.inc

Maybe they will decide to release it someday.

**Comment by oddMLan** [comment]
Very interesting. Couldn't be the same technique used here used for the Bloom filter? To avoid applying bloom to 2D elements such as text

**Comment by fzurita** [comment]
Yep, it definitely could be. The hardest part I see is coming up with the mask texture which I think could be done by rendering the 2D elements twice, first time to an FBO with a background color of black and drawing all 2D elements as white in the fragment shader. This will create the mask in the FBO. The 2nd time will be rendered to the regular frame buffer FBO.

**Comment by fzurita** [comment]
Looking back through old issues. Looks like the source code was released:

https://github.com/libretro/beetle-psx-libretro/blob/master/parallel-psx/glsl/quad.frag

**Comment by oddMLan** [comment]
This would be helpful to make FXAA not touch 2D elements. FXAA makes the text in Mario 64 look weird.
